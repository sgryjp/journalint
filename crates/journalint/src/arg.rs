use clap::Parser;

#[derive(Parser, Debug)]
#[command(author, version, about)]
pub struct Arguments {
    /// File to lint.
    pub filename: Option<String>,

    /// Fix found problems.
    #[arg(short, long)]
    pub fix: bool,

    /// Start as a language server.
    #[arg(long)]
    pub stdio: bool,
}
