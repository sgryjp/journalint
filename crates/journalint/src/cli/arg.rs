use clap::Parser;

use crate::cli::export::ExportFormat;
use crate::cli::report::ReportFormat;

#[derive(Parser, Debug)]
#[command(author, version, about)]
pub(crate) struct Arguments {
    /// File to lint.
    pub filename: Option<String>,

    /// Fix found problems.
    #[arg(short, long)]
    pub fix: bool,

    /// Report rule violations in the specified format.
    #[clap(value_enum)]
    #[arg(long, value_name = "FORMAT", default_value_t = ReportFormat::Fancy)]
    pub report: ReportFormat,

    /// Export journal entries in the specified format.
    #[arg(short, long, value_name = "FORMAT")]
    pub export: Option<ExportFormat>,

    /// Whether to extract activity prefix as codes on exporting.
    #[arg(long, default_value_t = false)]
    pub extract_activity_prefixes: bool,

    /// Start as a language server.
    #[arg(long)]
    pub stdio: bool,
}
