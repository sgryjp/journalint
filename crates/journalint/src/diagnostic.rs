use core::ops::Range;
use std::collections::HashMap;
use std::sync::Arc;

use lsp_types::DiagnosticSeverity;
use lsp_types::NumberOrString;
use lsp_types::TextEdit;
use lsp_types::Url;
use lsp_types::WorkspaceEdit;

use crate::code::Code;
use crate::linemap::LineMap;

static SOURCE_NAME: &str = "journalint";

/// Internal diagnostic data structure.
///
/// This is basically the same as lsp_types::Diagnostic except that this has a field
/// `span` of type Range<usize>, not a field `range` of type lsp_types::Range.
#[derive(Clone, Debug)]
pub struct Diagnostic {
    span: Range<usize>,
    code: Code,
    severity: DiagnosticSeverity,
    message: String,
    expectation: Option<String>,
    line_map: Arc<LineMap>,
}

impl Diagnostic {
    pub fn new_warning(
        span: Range<usize>,
        code: Code,
        message: String,
        expectation: Option<String>,
        line_mapper: Arc<LineMap>,
    ) -> Self {
        let severity = DiagnosticSeverity::WARNING;
        Self {
            span,
            code,
            severity,
            message,
            expectation,
            line_map: line_mapper,
        }
    }

    pub fn span(&self) -> &Range<usize> {
        &self.span
    }

    pub fn severity(&self) -> DiagnosticSeverity {
        self.severity
    }

    pub fn code(&self) -> &Code {
        &self.code
    }

    pub fn message(&self) -> &str {
        self.message.as_ref()
    }

    pub fn expectation(&self) -> Option<&String> {
        self.expectation.as_ref()
    }

    /// Try creating a WorkspaceEdit to fix the problem.
    pub fn fix(&self, url: &Url) -> Option<WorkspaceEdit> {
        // Create an edit data in the file to fix the issue
        let Some(new_text) = self.expectation() else {
            return None;
        };
        let edit = TextEdit::new(self.lsp_range(), new_text.clone());

        // Compose a "workspace edit" from it
        let edits = HashMap::from([(url.clone(), vec![edit])]);
        Some(WorkspaceEdit::new(edits))
    }

    // --- helper methods ---

    pub fn lsp_range(&self) -> lsp_types::Range {
        let start = self.line_map.position_from_offset(self.span.start);
        let end = self.line_map.position_from_offset(self.span.end);
        lsp_types::Range::new(start, end)
    }

    pub fn is_in_lsp_range(&self, range: &lsp_types::Range) -> bool {
        let start = self.line_map.offset_from_position(&range.start);
        let end = self.line_map.offset_from_position(&range.end);
        self.span.contains(&start) && self.span.contains(&end)
    }
}

impl From<Diagnostic> for lsp_types::Diagnostic {
    fn from(value: Diagnostic) -> Self {
        let code = value.code().as_str().to_string();
        let range = lsp_types::Range::new(
            value.line_map.position_from_offset(value.span().start),
            value.line_map.position_from_offset(value.span().end),
        );
        lsp_types::Diagnostic::new(
            range,
            Some(value.severity()),
            Some(NumberOrString::String(code)),
            Some(SOURCE_NAME.to_string()),
            value.message().to_owned(),
            None,
            None,
        )
    }
}
