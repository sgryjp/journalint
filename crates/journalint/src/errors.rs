use crossbeam_channel::SendError;
use journalint_parse::errors::UnknownViolationCode;
use lsp_server::Message;
use lsp_server::ProtocolError;
use lsp_types::Url;
use thiserror::Error;

use journalint_parse::errors::InvalidTimeValueError;

// ----------------------------------------------------------------------------

/// Error for CLI main function.
#[derive(Default)]
pub struct CliError {
    exit_status: i32,
    message: Option<String>,
}

impl CliError {
    pub fn new(exit_status: i32) -> Self {
        Self {
            exit_status,
            message: None,
        }
    }

    pub fn with_message(&self, message: String) -> Self {
        Self {
            message: Some(message),
            ..*self
        }
    }

    pub fn exit_status(&self) -> i32 {
        self.exit_status
    }

    pub fn message(&self) -> Option<&String> {
        self.message.as_ref()
    }
}

// ----------------------------------------------------------------------------

#[derive(Error, Debug)]
pub enum JournalintError {
    #[error("UNEXPECTED ERROR: {0}")]
    UnexpectedError(String),

    #[error("{}", source)]
    UnknownViolationCode {
        #[from]
        source: UnknownViolationCode,
    },

    #[error("Unknown command: {0}")]
    UnknownCommand(String),

    #[error("Unexpected arguments: {0}")]
    UnexpectedArguments(String),

    #[error("Unsupported URL: {url}")]
    UnsupportedUrl { url: Url },

    #[error("Doucument not found: {url}")]
    DocumentNotFound { url: Url },

    #[error("Parse error: {}", .source)]
    InvalidTimeValueError {
        #[from]
        source: InvalidTimeValueError,
    },

    #[error("Required value is missing: {name}")]
    MissingRequiredValue { name: String },

    #[error("Target not found for command '{command}'")]
    CommandTargetNotFound { command: String },

    #[error("LSP communication error: {0}")]
    LspCommunicationError(String),

    #[error("I/O Error: {}", .source)]
    Io {
        #[from]
        source: std::io::Error,
    },

    #[error("Serialization error: {}", .source)]
    SerializationError {
        #[from]
        source: serde_json::error::Error,
    },

    #[error("Parsing date or time failed: {}", .source)]
    ChronoParseError {
        #[from]
        source: chrono::ParseError,
    },
}

impl From<SendError<Message>> for JournalintError {
    fn from(value: SendError<Message>) -> Self {
        JournalintError::LspCommunicationError(value.to_string())
    }
}

impl From<ProtocolError> for JournalintError {
    fn from(value: ProtocolError) -> Self {
        JournalintError::LspCommunicationError(value.to_string())
    }
}
