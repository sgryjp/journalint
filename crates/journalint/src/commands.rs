use std::fs::write;
use std::path::Path;

use lsp_types::{Url, WorkspaceEdit};

use crate::code::Code;
use crate::diagnostic::Diagnostic;
use crate::errors::JournalintError;
use crate::service::ServerState;

pub const RECALCULATE_DURATION: &str = "journalint.recalculateDuration";
pub const REPLACE_WITH_PREVIOUS_END_TIME: &str = "journalint.replaceWithPreviousEndTime";
pub const ALL_COMMANDS: [&str; 2] = [RECALCULATE_DURATION, REPLACE_WITH_PREVIOUS_END_TIME];

pub trait Command {
    fn title(&self) -> &str;
    fn command(&self) -> &str;
    fn execute(
        &self,
        state: &ServerState,
        url: &Url,
        range: &lsp_types::Range,
    ) -> Option<WorkspaceEdit>;
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct CommandParams {
    url: Url,
    range: lsp_types::Range,
}

// -----------------------------------------------------------------------------
#[derive(Debug)]
pub struct RecalculateDuration {}

impl Command for RecalculateDuration {
    fn title(&self) -> &str {
        "Recalculate duration by the interval between start and end time"
    }

    fn command(&self) -> &str {
        RECALCULATE_DURATION
    }

    fn execute(
        &self,
        state: &ServerState,
        url: &Url,
        range: &lsp_types::Range,
    ) -> Option<WorkspaceEdit> {
        state
            .find_diagnostic(url, range, Code::IncorrectDuration)
            .and_then(|d| d.fix(url))
    }
}

// -----------------------------------------------------------------------------
#[derive(Debug)]
pub struct ReplaceWithPreviousEndTime {}

impl Command for ReplaceWithPreviousEndTime {
    fn title(&self) -> &str {
        "Replace with the previous entry's end time"
    }

    fn command(&self) -> &str {
        REPLACE_WITH_PREVIOUS_END_TIME
    }

    fn execute(
        &self,
        state: &ServerState,
        url: &Url,
        range: &lsp_types::Range,
    ) -> Option<WorkspaceEdit> {
        state
            .find_diagnostic(url, range, Code::TimeJumped)
            .and_then(|d| d.fix(url))
    }
}

// -----------------------------------------------------------------------------
pub fn list_available_code_actions(code: &Code) -> Option<Vec<Box<dyn Command>>> {
    match code {
        Code::ParseError => None,
        Code::MismatchedDates => None,
        Code::InvalidStartTime => None,
        Code::InvalidEndTime => None,
        Code::MissingDate => None,
        Code::MissingStartTime => None,
        Code::MissingEndTime => None,
        Code::TimeJumped => Some(vec![Box::new(ReplaceWithPreviousEndTime {})]),
        Code::NegativeTimeRange => None,
        Code::IncorrectDuration => Some(vec![Box::new(RecalculateDuration {})]),
    }
}

pub fn get_command_by_name(name: &str) -> Option<Box<dyn Command>> {
    match name {
        RECALCULATE_DURATION => Some(Box::new(RecalculateDuration {})),
        REPLACE_WITH_PREVIOUS_END_TIME => Some(Box::new(ReplaceWithPreviousEndTime {})),
        _ => None,
    }
}

pub fn fix(diagnostic: &Diagnostic, content: &str, path: &Path) -> Result<(), JournalintError> {
    // TODO: Move somewhere else
    let span = diagnostic.span();
    let (start, end) = (span.start, span.end);

    if let Some(expectation) = diagnostic.expectation() {
        let mut buf = String::with_capacity(content.len());
        buf.push_str(&content[..start]);
        buf.push_str(expectation.as_str());
        buf.push_str(&content[end..]);
        write(path, buf)?;
    };
    Ok(())
}
