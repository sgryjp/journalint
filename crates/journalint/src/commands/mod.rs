//! This module provides commands of journalint language server.
mod autofix;
mod recalculate_duration;
mod replace_with_previous_end_time;
mod use_date_in_filename_visitor;

use std::fs::write;
use std::path::Path;

use lsp_types::{Url, WorkspaceEdit};

use crate::code::Code;
pub use crate::commands::autofix::AutofixCommand;
use crate::diagnostic::Diagnostic;
use crate::errors::JournalintError;
use crate::service::ServerState;

/// Command of journalint.
///
/// Currently I only think of auto-fix commands so this must be unsuitable for other kinds of
/// commands.
pub trait Command {
    /// Get short description of this command which is meant to be used in UI.
    fn title(&self) -> &str;

    /// Get machine-readable identifier of this command.
    fn id(&self) -> &str;

    /// Get a diagnostic code which is fixable by this command.
    fn fixable_codes(&self) -> Code;

    /// Executes this command.
    ///
    /// In case of fix commands, the result is the change set to be applied to the document.
    /// Note that it will be `Ok(None)` if there is nothing to do.
    fn execute(
        &self,
        state: &ServerState,
        url: &Url,
        range: &lsp_types::Range,
    ) -> Result<Option<WorkspaceEdit>, JournalintError>;
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct CommandParams {
    url: Url,
    range: lsp_types::Range,
}

// -----------------------------------------------------------------------------

pub fn fix(diagnostic: &Diagnostic, content: &str, path: &Path) -> Result<(), JournalintError> {
    // TODO: Move somewhere else
    let span = diagnostic.span();
    let (start, end) = (span.start, span.end);

    if let Some(expectation) = diagnostic.expectation() {
        let mut buf = String::with_capacity(content.len());
        buf.push_str(&content[..start]);
        buf.push_str(expectation.as_str());
        buf.push_str(&content[end..]);
        write(path, buf)?;
    };
    Ok(())
}
