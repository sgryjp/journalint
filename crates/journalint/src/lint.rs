//! Provides lint logic.
//!
//! See module `ast` for AST related features, and module `parse` for parsing logic.
use std::ops::Range;
use std::path::PathBuf;
use std::sync::Arc;
use std::time::Duration;

use chrono::{DateTime, NaiveDate, Timelike, Utc};
use lsp_types::Url;

use crate::ast::{walk, Expr, LooseTime, Visitor};
use crate::code::Code;
use crate::diagnostic::{Diagnostic, DiagnosticRelatedInformation};
use crate::errors::JournalintError;
use crate::linemap::LineMap;

pub struct Linter<'a> {
    source: &'a Url,
    diagnostics: Vec<Diagnostic>,
    line_map: Arc<LineMap>,

    fm_date: Option<(NaiveDate, Range<usize>)>,
    fm_start: Option<(LooseTime, Range<usize>)>,
    fm_start_datetime: Option<DateTime<Utc>>,
    fm_end: Option<(LooseTime, Range<usize>)>,
    fm_end_datetime: Option<DateTime<Utc>>,

    entry_start: Option<(DateTime<Utc>, Range<usize>)>,
    entry_end: Option<(DateTime<Utc>, Range<usize>)>,
    prev_entry_end: Option<(DateTime<Utc>, Range<usize>)>,
}

impl<'a> Linter<'a> {
    pub fn new(source: &Url, line_map: Arc<LineMap>) -> Linter {
        Linter {
            source,
            diagnostics: vec![],
            line_map,

            fm_date: None,
            fm_start: None,
            fm_start_datetime: None,
            fm_end: None,
            fm_end_datetime: None,

            entry_start: None,
            entry_end: None,
            prev_entry_end: None,
        }
    }

    // Check if date value in the front matter does not match the one in the filename
    fn check_fm_date_matches_filename(&mut self, value: &NaiveDate, span: &Range<usize>) {
        if let Some(stem) = PathBuf::from(self.source.path())
            .file_stem()
            .and_then(|s| s.to_str())
        {
            if let Ok(date_in_filename) = NaiveDate::parse_from_str(stem, "%Y-%m-%d") {
                if date_in_filename != *value {
                    let expectation = date_in_filename.format("%Y-%m-%d").to_string();
                    self.diagnostics.push(Diagnostic::new_warning(
                        span.clone(),
                        Code::MismatchedDates,
                        format!(
                            "Date is different from the one in the filename: expected to be {}",
                            expectation.as_str()
                        ),
                        Some(expectation),
                        None,
                        self.line_map.clone(),
                    ));
                }
            }
        }
    }

    fn check_fm_date_exists(&mut self, span: &Range<usize>) {
        if self.fm_date.is_none() {
            self.diagnostics.push(Diagnostic::new_warning(
                span.clone(),
                Code::MissingDate,
                "Field 'date' is missing".to_string(),
                None,
                None,
                self.line_map.clone(),
            ));
        }
    }

    fn check_fm_start_exists(&mut self, span: &Range<usize>) {
        if self.fm_start.is_none() {
            self.diagnostics.push(Diagnostic::new_warning(
                span.clone(),
                Code::MissingStartTime,
                "Field 'start' is missing".to_string(),
                None,
                None,
                self.line_map.clone(),
            ));
        }
    }

    fn check_fm_start_is_valid(&mut self) -> Option<DateTime<Utc>> {
        let Some((date, _)) = self.fm_date.as_ref() else {
            return None;
        };
        let Some((start, start_span)) = self.fm_start.as_ref() else {
            return None;
        };

        match start.to_datetime(*date) {
            Ok(dt) => Some(dt),
            Err(e) => {
                self.diagnostics.push(Diagnostic::new_warning(
                    start_span.clone(),
                    Code::InvalidStartTime,
                    format!("Invalid start time: {e}"),
                    None,
                    None,
                    self.line_map.clone(),
                ));
                None
            }
        }
    }

    fn check_fm_end_exists(&mut self, span: &Range<usize>) {
        if self.fm_end.is_none() {
            self.diagnostics.push(Diagnostic::new_warning(
                span.clone(),
                Code::MissingEndTime,
                "Field 'end' is missing".to_string(),
                None,
                None,
                self.line_map.clone(),
            ));
        }
    }

    fn check_fm_end_is_valid(&mut self) -> Option<DateTime<Utc>> {
        let Some((date, _)) = self.fm_date.as_ref() else {
            return None;
        };
        let Some((end, end_span)) = self.fm_end.as_ref() else {
            return None;
        };

        match end.to_datetime(*date) {
            Ok(dt) => Some(dt),
            Err(e) => {
                self.diagnostics.push(Diagnostic::new_warning(
                    end_span.clone(),
                    Code::InvalidEndTime,
                    format!("Invalid end time: {e}"),
                    None,
                    None,
                    self.line_map.clone(),
                ));
                None
            }
        }
    }

    // Check if start time matches the end of the previous entry
    fn check_prev_end_equals_next_start(&mut self, start_dt: DateTime<Utc>, span: &Range<usize>) {
        if let Some((prev_end_dt, prev_end_range)) = self.prev_entry_end.as_ref() {
            if start_dt != *prev_end_dt {
                let expectation = prev_end_dt.format("%H:%M").to_string();
                self.diagnostics.push(Diagnostic::new_warning(
                    span.clone(),
                    Code::TimeJumped,
                    format!("The start time does not match the previous entry's end time, which is {expectation}"),
                    Some(expectation),
                    Some(vec![DiagnosticRelatedInformation::new(
                        self.source.clone(),
                        prev_end_range.clone(),
                        format!(
                            "Previous entry's end time is {:02}:{:02}",
                            prev_end_dt.hour(),
                            prev_end_dt.minute()
                        ),
                        self.line_map.clone(),
                    )]),
                    self.line_map.clone(),
                ));
            }
        }
    }

    fn check_start_time(
        &mut self,
        value: &LooseTime,
        span: &Range<usize>,
    ) -> Option<DateTime<Utc>> {
        let Some((date, _)) = self.fm_date else {
            return None;
        };

        match value.to_datetime(date) {
            Ok(dt) => Some(dt),
            Err(e) => {
                // Start time is not a valid value
                self.diagnostics.push(Diagnostic::new_warning(
                    span.clone(),
                    Code::InvalidStartTime,
                    format!("Invalid start time: {e}"),
                    None,
                    None,
                    self.line_map.clone(),
                ));
                None
            }
        }
    }

    fn check_end_time(&mut self, value: &LooseTime, span: &Range<usize>) -> Option<DateTime<Utc>> {
        let Some((date, _)) = self.fm_date else {
            return None;
        };

        match value.to_datetime(date) {
            Ok(dt) => Some(dt),
            Err(e) => {
                self.diagnostics.push(Diagnostic::new_warning(
                    span.clone(),
                    Code::InvalidEndTime,
                    format!("Invalid end time: {e}"),
                    None,
                    None,
                    self.line_map.clone(),
                ));
                None
            }
        }
    }

    fn check_end_time_exceeds_start_time(&mut self) {
        let Some((start, _)) = self.entry_start.as_ref() else {
            return;
        };
        let Some((end, end_span)) = self.entry_end.as_ref() else {
            return;
        };

        let Ok(_) = (*end - *start).to_std() else {
            self.diagnostics.push(Diagnostic::new_warning(
                end_span.clone(),
                Code::NegativeTimeRange,
                format!(
                    "End time is not ahead of start time ({})",
                    start.format("%H:%M"),
                ),
                None,
                None,
                self.line_map.clone(),
            ));
            return;
        };
    }

    fn check_duration_matches_end_minus_start(&mut self, value: &Duration, span: &Range<usize>) {
        let Some((start, _)) = self.entry_start.as_ref() else {
            return;
        };
        let Some((end, _)) = self.entry_end.as_ref() else {
            return;
        };
        let Ok(calculated) = (*end - *start).to_std() else {
            return;
        };
        let written = value;
        if calculated != *written {
            let expectation = calculated.as_secs_f64() / 3600.0;
            let expectation = format!("{expectation:1.2}");
            self.diagnostics.push(Diagnostic::new_warning(
                span.clone(),
                Code::IncorrectDuration,
                format!("Incorrect duration: expected {expectation}"),
                Some(expectation),
                None,
                self.line_map.clone(),
            ));
        }
    }
}

impl Visitor for Linter<'_> {
    fn on_visit_fm_date(
        &mut self,
        value: &NaiveDate,
        span: &Range<usize>,
    ) -> Result<(), JournalintError> {
        self.fm_date = Some((*value, span.clone()));
        self.check_fm_date_matches_filename(value, span);
        Ok(())
    }

    fn on_visit_fm_start(
        &mut self,
        value: &LooseTime,
        span: &Range<usize>,
    ) -> Result<(), JournalintError> {
        self.fm_start = Some((value.clone(), span.clone()));
        Ok(())
    }

    fn on_visit_fm_end(
        &mut self,
        value: &LooseTime,
        span: &Range<usize>,
    ) -> Result<(), JournalintError> {
        self.fm_end = Some((value.clone(), span.clone()));
        Ok(())
    }

    fn on_leave_fm(&mut self, span: &Range<usize>) -> Result<(), JournalintError> {
        // Calculate exact time of start and end
        self.fm_start_datetime = self.check_fm_start_is_valid();
        self.fm_end_datetime = self.check_fm_end_is_valid();

        // Warn if one of date, start and end is missing
        self.check_fm_date_exists(span);
        self.check_fm_start_exists(span);
        self.check_fm_end_exists(span);

        Ok(())
    }

    fn on_visit_start_time(
        &mut self,
        value: &LooseTime,
        span: &Range<usize>,
    ) -> Result<(), JournalintError> {
        if let Some(start_dt) = self.check_start_time(value, span) {
            self.entry_start = Some((start_dt, span.clone()));
            self.check_prev_end_equals_next_start(start_dt, span);
        }
        Ok(())
    }

    fn on_visit_end_time(
        &mut self,
        value: &LooseTime,
        span: &Range<usize>,
    ) -> Result<(), JournalintError> {
        if let Some(dt) = self.check_end_time(value, span) {
            self.entry_end = Some((dt, span.clone()));
        }
        Ok(())
    }

    fn on_visit_duration(
        &mut self,
        value: &Duration,
        span: &Range<usize>,
    ) -> Result<(), JournalintError> {
        self.check_end_time_exceeds_start_time();
        self.check_duration_matches_end_minus_start(value, span);
        Ok(())
    }
}

pub fn lint(
    journal: &Expr,
    url: &Url,
    line_map: Arc<LineMap>,
) -> Result<Vec<Diagnostic>, JournalintError> {
    let mut visitor = Linter::new(url, line_map);
    walk(journal, &mut visitor)?;
    Ok(visitor.diagnostics)
}
