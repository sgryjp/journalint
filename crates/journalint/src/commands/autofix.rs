//! Autofix commands
use std::collections::HashMap;

use lsp_types::{TextEdit, Url, WorkspaceEdit};
use strum::EnumIter;

use crate::code::Code;
use crate::commands::Command;
use crate::errors::JournalintError;
use crate::service::ServerState;

use super::use_date_in_filename_visitor;

/// Auto-fix command.
#[derive(Debug, EnumIter)]
pub enum AutofixCommand {
    RecalculateDuration,
    ReplaceWithPreviousEndTime,
    UseDateInFilename,
}

impl Command for AutofixCommand {
    fn title(&self) -> &str {
        match self {
            AutofixCommand::RecalculateDuration => {
                "Recalculate duration by the interval between start and end time"
            }
            AutofixCommand::ReplaceWithPreviousEndTime => {
                "Replace with the previous entry's end time"
            }
            AutofixCommand::UseDateInFilename => "Use date embedded in the filename",
        }
    }

    fn id(&self) -> &str {
        match self {
            AutofixCommand::RecalculateDuration => "journalint.recalculateDuration",
            AutofixCommand::ReplaceWithPreviousEndTime => "journalint.replaceWithPreviousEndTime",
            AutofixCommand::UseDateInFilename => "journalint.useDateInFilename",
        }
    }

    fn fixable_codes(&self) -> Code {
        match self {
            AutofixCommand::RecalculateDuration => Code::IncorrectDuration,
            AutofixCommand::ReplaceWithPreviousEndTime => Code::TimeJumped,
            AutofixCommand::UseDateInFilename => Code::MismatchedDates,
        }
    }

    /// Execute an auto-fix command.
    ///
    /// # Arguments
    ///
    /// * `state` - State of the language server
    /// * `url` - URL of the document
    /// * `range` - Range of the selection at the time this command was invoked.
    fn execute(
        &self,
        state: &ServerState,
        url: &Url,
        range: &lsp_types::Range,
    ) -> Result<Option<WorkspaceEdit>, JournalintError> {
        match self {
            AutofixCommand::RecalculateDuration => execute_fix(self, state, url, range),
            AutofixCommand::ReplaceWithPreviousEndTime => execute_fix(self, state, url, range),
            AutofixCommand::UseDateInFilename => {
                // Get state of the document
                let doc_state = state.document_state(url)?;
                let line_map = doc_state.line_map();
                let ast = doc_state.ast().ok_or_else(|| {
                    JournalintError::UnexpectedError(format!(
                        "No AST available for the document: {url}"
                    ))
                })?;
                use_date_in_filename_visitor::execute(url, line_map, ast)
            }
        }
    }
}

fn execute_fix(
    command: &dyn Command,
    state: &ServerState,
    url: &Url,
    range: &lsp_types::Range,
) -> Result<Option<WorkspaceEdit>, JournalintError> {
    let doc_state = state.document_state(url)?;

    // Find matching diagnostic object
    let code = command.fixable_codes();
    let diagnostic = state
        .find_diagnostic(url, range, &code)
        .ok_or_else(|| JournalintError::UnexpectedError(format!(
            "No corresponding diagnostic found to fix: {{command: {}, url: {}, range: {:?}, code: {}}}",
            command.id(), url, range, code
        )))?;
    let range_to_replace = doc_state.line_map().span_to_lsp_range(diagnostic.span());

    // Create an edit data in the file to fix the issue
    let Some(new_text) = diagnostic.expectation() else {
        return Ok(None);
    };
    let edit = TextEdit::new(range_to_replace, new_text.clone());

    // Compose a "workspace edit" from it
    let edits = HashMap::from([(url.clone(), vec![edit])]);
    Ok(Some(WorkspaceEdit::new(edits)))
}
