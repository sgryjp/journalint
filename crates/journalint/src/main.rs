mod arg;
mod ast;
mod code;
mod commands;
mod diagnostic;
mod errors;
mod export;
mod linemap;
mod lint;
mod parse;
mod service;
mod textedit;

use std::env;
use std::fs::read_to_string;
use std::path::PathBuf;

use ariadne::Color;
use ariadne::Label;
use ariadne::Report;
use ariadne::ReportKind;
use ariadne::Source;
use clap::Parser;
use commands::apply_text_edit;
use commands::Command;
use diagnostic::Diagnostic;
use env_logger::TimestampPrecision;
use errors::CliError;
use log::error;
use lsp_types::Url;

use crate::arg::Arguments;
use crate::errors::JournalintError;
use crate::lint::lint;
use crate::parse::parse;

const E_UNEXPECTED: exitcode::ExitCode = 1;

/// Entry point of journalint CLI.
fn main() -> Result<(), JournalintError> {
    // Initialize logging
    env_logger::builder()
        .format_timestamp(Some(TimestampPrecision::Millis))
        .init();

    // Parse arguments
    let args = Arguments::parse_from(env::args());

    // Start the service or the CLI
    if args.stdio {
        service::main()
    } else {
        let exit_status = match cli_main(args) {
            Ok(()) => exitcode::OK,
            Err(e) => {
                if let Some(msg) = e.message() {
                    error!("{}", msg);
                };
                e.exit_status()
            }
        };
        std::process::exit(exit_status);
    }
}

fn cli_main(args: Arguments) -> Result<(), CliError> {
    // Make sure a filename was given
    let filename = args.filename.ok_or(
        CliError::new(exitcode::USAGE).with_message("FILENAME must be specified.".to_string()),
    )?;
    let path = PathBuf::from(&filename).canonicalize().map_err(|e| {
        CliError::new(exitcode::IOERR).with_message(format!(
            "Failed to canonicalize the filename {filename:?}: {e:?}"
        ))
    })?;
    let url = Url::from_file_path(path.clone()).map_err(|_| {
        CliError::new(E_UNEXPECTED)
            .with_message(format!("Failed to compose URL from path {:?}", &path))
    })?;

    // Load the content
    let content = read_to_string(&path).map_err(|e| {
        CliError::new(exitcode::IOERR).with_message(format!("Failed to read {filename:?}: {e:?}"))
    })?;

    // Parse the content and lint the AST unless parsing itself failed
    let (journal, mut diagnostics, line_map) = parse(&content);
    if let Some(journal) = journal.as_ref() {
        let mut d = lint(journal, &url, line_map.clone()).map_err(|e| {
            CliError::new(E_UNEXPECTED).with_message(format!("Failed on linting: {e:?}"))
        })?;
        diagnostics.append(&mut d);
    }

    // Execute specified task against the AST and diagnostics
    if args.fix {
        // Sort diagnostics in reverse order
        diagnostics.sort_by(|a, b| b.span().start.cmp(&a.span().start));

        // Fix one by one
        for d in diagnostics.iter().as_ref() {
            // Check if there is a default auto-fix command for the diagnostic.
            let (Some(ast_root), Some(command)) = (&journal, d.code().default_autofix()) else {
                continue; // unavailable
            };

            // Execute the default auto-fix command.
            let text_edit = command
                .execute(&url, ast_root, d.span())
                .map_err(|e| CliError::new(E_UNEXPECTED).with_message(e.to_string()))?;
            if let Some(text_edit) = text_edit {
                apply_text_edit(&url, text_edit)
                    .map_err(|e| CliError::new(E_UNEXPECTED).with_message(e.to_string()))?;
            }
        }
    } else {
        // Write diagnostic report to stderr
        diagnostics
            .iter()
            .for_each(|d| report(&content, Some(&filename), d));

        // Export parsed data to stdout
        if let Some(fmt) = args.export {
            if let Some(journal) = journal {
                let mut writer = std::io::stdout();
                crate::export::export(fmt, journal, &mut writer).map_err(|e| {
                    CliError::new(3).with_message(format!("Failed to export data: {:?}", e))
                })?;
            }
        }
    }

    Ok(())
}

/// Write a human readable report of a diagnostic
fn report(content: &str, filename: Option<&str>, diagnostic: &Diagnostic) {
    let stdin_source_name = "<STDIN>".to_string();
    let filename = filename.unwrap_or(&stdin_source_name);
    let start = diagnostic.span().start;
    let end = diagnostic.span().end;
    let message = diagnostic.message();

    Report::build(ReportKind::Error, filename, start)
        .with_message(message)
        .with_label(
            Label::new((filename, start..end))
                .with_color(Color::Red)
                .with_message(message),
        )
        .finish()
        .eprint((filename, Source::from(content)))
        .unwrap();
}

#[cfg(test)]
mod snapshot_tests {
    use std::{ffi::OsStr, fs};

    use super::*;

    fn parse_and_lint(url: &Url, content: &str) -> Vec<Diagnostic> {
        let (journal, mut diagnostics, line_map) = parse(&content);
        if let Some(journal) = journal {
            let mut d = lint(&journal, &url, line_map).expect("FAILED TO LINT");
            diagnostics.append(&mut d);
        };

        diagnostics
    }

    #[test]
    fn test() {
        for entry in fs::read_dir("src/snapshots").unwrap() {
            let path = entry.and_then(|ent| ent.path().canonicalize()).unwrap();
            let path = path.as_path();
            if path.extension() != Some(OsStr::new("md")) {
                continue;
            }
            let url = &Url::from_file_path(path)
                .expect(&format!("failed to compose a URL from path: {:?}", path));
            let content = match read_to_string(path) {
                Ok(content) => content,
                Err(err) => panic!("failed to read a file: {{path: {:?}, err:{}}}", path, err),
            };

            let diagnostics = parse_and_lint(&url, &content)
                .iter()
                .map(|d| d.clone().into())
                .collect::<Vec<lsp_types::Diagnostic>>();
            insta::assert_yaml_snapshot!(path.file_stem().unwrap().to_str().unwrap(), diagnostics);
        }
    }
}
