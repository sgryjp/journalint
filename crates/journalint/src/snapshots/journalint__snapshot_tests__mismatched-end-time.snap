---
source: crates/journalint/src/main.rs
expression: diagnostics
---
- range:
    start:
      line: 3
      character: 5
    end:
      line: 3
      character: 10
  severity: 2
  code: endtime-mismatch
  source: journalint
  message: "End time in the front-matter is different from the one of the last entry: expected to be 10:00."
  relatedInformation:
    - location:
        uri: "file:///snapshots/mismatched-end-time.md"
        range:
          start:
            line: 8
            character: 8
          end:
            line: 8
            character: 13
      message: "The last entry ends with 10:00."
